### DISCUSSION TITLE
Object detection on the streams

### DISCUSSION THREAD

**Discussion by rpatidar** [bluenviron/mediamtx#1478] [Questions] [answered]
I have a usecase where we need to run object detection on the streams, this will be multiple streams in thousands.
1. Publish the raw stream via webrtc
2. modify the stream inject object detection portion and also publish via webrtc 

I will have one api server where the frame can be sent for the detection of object. 
Any suggestion how can invoke API in between and again stream it back to be used as part of webrtc etc  as new stream.

1. Overall i would like to keep the CPU overhead to minimal reducing the intermediates encoding/decoding as much as possible.
2. Reduce coupling or minimal changes to rtsp-simple-server so can invoke the object detection api server independently and run some forking. and send it back to rtsp-simple server to publish it back to some new webrtc endpoints.

question :
1. does listening the local output stream  from the simple rtsp-server and decoding it again is a good idea, it will lead to double decoding of rtsp stream ? 

sample illustration of problem.
```
                                         
                 ┌────────────┐
                 │            │
                 │  Object    │
                 │  Detection │
                 │            │
                 └──▲────┬────┘
                    │    │
                    │    │
                    │    │                ┌─────────┐
                  ┌─┴────▼──┐             │ WEB RTC │
                  │  RTSP   ├────────────►│ Viewer1 │
──RTSP Stream────►│  Simple │             └─────────┘
                  │  Server ├───────┐
                  └─────────┘       │     ┌─────────┐
                                    │     │Object St│
                                    └────►│WEB RTC V2
                                          └──────────
```

**Comment by rpatidar** [comment, marked as answer]
I will be going with polling the streaming and inserting back after object detection way.

**Comment by aler9** [comment]
Hello, the only element of your scheme that needs to decode the video (beside the viewers) is the object detection block, while the server doesn't decode the stream and there's no double decoding.

My suggestion is to make the object detection microservice read the stream in RTSP/H264 format and perform video decoding directly on the service, rather than sending it frames through an API. Video codecs are actually the only feasible way to transfer full HD video streams through copper cable, therefore you shouldn't decode the video before the stream arrives on its destination.

Another suggestion is to the GPU whenever possible, when decoding the video, when performing object detection and when re-encoding the video, that's the only way to run such processes without saturating the CPU.

GStreamer and Deepstream allow to perform all parts of the process.

[a comment by an automated account is not shown]
